### PR TITLE
make public partial evaluation API experimental

### PR DESCRIPTION
**Pull request by andrewmwells-amazon**
make public partial evaluation API experimental

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [ ] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).
- [ ] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).
- [x] A bug fix or other functionality change requiring a patch to `cedar-policy`.
- [ ] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)
- [ ] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).
- [ ] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar Dafny model or DRT infrastructure.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)
- [ ] Requires updates, but I do not plan to make them in the near future. (Make sure that your changes are hidden behind a feature flag to mark them as experimental.)
- [ ] I'm not sure how my change impacts `cedar-spec`. (Post your PR anyways, and we'll discuss in the comments.)

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by shaobo-he-aws — APPROVED** [review]
LGTM.

File: cedar-policy/src/api.rs [hunk 1]
```diff
@@ -301,6 +301,7 @@ impl Authorizer {
         self.0.is_authorized(&r.0, &p.ast, &e.0).into()
     }
 
+    #[cfg(feature = "partial_eval")]
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:304]: Typically this annotation would go below the `///` docs, not above them

**Review by cdisselkoen — CHANGES_REQUESTED** [review]
I feel like we should do slightly more in this PR. 
- move `Entities::partial()` behind this flag
- move `PartialResponse` behind this flag
- move `ResidualResponse` behind this flag

**Comment by andrewmwells-amazon** [comment]
I think we want to make sure the build with / without `--features "experimental"` has no warnings? That's what I've done in the latest commit. The downside is it requires `#[allow(unused)]` annotations we may then forget to remove if/when the feature is no longer experimental. I added comments naming the feature to each usage; maybe that's good enough.

File: cedar-policy/Cargo.toml [hunk 2]
```diff
@@ -37,6 +37,10 @@ corpus-timing = []
 
 integration_testing = []
 
+# Experimental features
+experimental = ["partial-eval"]
```
Comment (cdisselkoen) [inline cedar-policy/Cargo.toml:41]: Is the intent that `experimental` would turn on all experimental features?  If so can we have a comment documenting that?

File: cedar-policy/src/api.rs [hunk 3]
```diff
@@ -1507,6 +1509,7 @@ impl Template {
     /// create the EST from the policy text or CST instead, as the conversion
     /// to AST is lossy. ESTs generated by this method will reflect the AST and
     /// not the original policy syntax.
+    #[allow(unused)] //Feature: partial-eval
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:1512]: This could be `#[cfg_attr(feature = "partial-eval", allow(unused))]`

Likewise elsewhere

**Comment by andrewmwells-amazon** [comment]
> * move `PartialResponse` behind this flag

Hmm, this change will be more invasive. Specifically it would require rewriting `is_authorized_core(`.

**Comment by andrewmwells-amazon** [comment]
> I feel like we should do slightly more in this PR.
> 
> * move `Entities::partial()` behind this flag
> * move `ResidualResponse` behind this flag

Did these two.

**Comment by andrewmwells-amazon** [comment]
> I think we want to make sure the build with / without `--features "experimental"` has no warnings? That's what I've done in the latest commit. The downside is it requires `#[allow(unused)]` annotations we may then forget to remove if/when the feature is no longer experimental. I added comments naming the feature to each usage; maybe that's good enough.

`#[cfg_attr(not(feature = "partial-eval"), allow(unused))]` works instead

**Review by shaobo-he-aws — APPROVED** [review]

File: cedar-policy/src/api.rs [hunk 4]
```diff
@@ -2338,6 +2346,7 @@ pub fn eval_expression(
 }
 
 #[cfg(test)]
+#[cfg(feature = "partial-eval")]
 mod test {
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:2350]: could we maybe rename the `test` module to be clear it's only tests related to partial evaluation?  Doesn't have to block this PR though

**Review by cdisselkoen — APPROVED** [review]

**Comment by andrewmwells-amazon** [comment]
>Hmm, this change will be more invasive. Specifically it would require rewriting `is_authorized_core(`

Did the more granular change I think Craig intended instead of the more invasive change of hiding just the `Residual(ResidualResponse)` variant.

**Review by aaronjeline — APPROVED** [review]

**Pull request merged**